contracts/meta-staking/src/contract.rs

```diff
@@ -1,55 +1,417 @@
+        // Check this is a consumer calling this, fails if no consumer loads
+        CONSUMERS.has(deps.storage, &info.sender);
+
+        // TODO Need to figure out how many rewards we got, so can send them
+        // to the consumer contract
+
+        // Withdraw rewards as a submessage
+        let withdraw_msg = SubMsg::reply_on_success(
+            DistributionMsg::WithdrawDelegatorReward { validator },
+            WITHDRAW_REWARDS_REPLY_ID,
+        );
+
+        // TODO On reply, send funds to consumer contract
+
+        Ok(Response::default().add_submessage(withdraw_msg))
+    }
+}
+
+// TODO query this info by consumer...
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::AllDelegations { delegator } => query::all_delegations(deps, delegator),
+        QueryMsg::AllValidators {} => query::all_validators(deps),
+        QueryMsg::BondedDenom {} => query::bonded_denom(deps),
+        QueryMsg::Config {} => query::config(deps),
+        QueryMsg::Delegation {
+            delegator,
+            validator,
+        } => query::delegation(deps, delegator, validator),
+        QueryMsg::Validator { address } => query::validator(deps, address),
+    }
+}
+
+mod query {
+    use cosmwasm_std::to_binary;
+
+    use super::*;
+
+    pub fn all_delegations(deps: Deps, delegator: String) -> StdResult<Binary> {
+        let all_delegations = deps.querier.query_all_delegations(delegator)?;
+        to_binary(&all_delegations)
+    }
+
+    pub fn all_validators(deps: Deps) -> StdResult<Binary> {
+        let all_validators = deps.querier.query_all_validators()?;
+        to_binary(&all_validators)
+    }
+
+    pub fn bonded_denom(deps: Deps) -> StdResult<Binary> {
+        let bonded_denom = deps.querier.query_bonded_denom()?;
+        to_binary(&bonded_denom)
+    }
+
+    pub fn delegation(deps: Deps, delegator: String, validator: String) -> StdResult<Binary> {
+        let delegation = deps.querier.query_delegation(delegator, validator)?;
+        to_binary(&delegation)
+    }
+
+    pub fn validator(deps: Deps, address: String) -> StdResult<Binary> {
+        let validator = deps.querier.query_validator(address)?;
+        to_binary(&validator)
+    }
+
+    pub fn config(deps: Deps) -> StdResult<Binary> {
+        let config = CONFIG.load(deps.storage)?;
+        to_binary(&config)
+    }
 }
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn query(_deps: Deps, _env: Env, _msg: QueryMsg) -> StdResult<Binary> {
-    unimplemented!();
+pub fn sudo(deps: DepsMut, env: Env, msg: SudoMsg) -> Result<Response, ContractError> {
+    match msg {
+        SudoMsg::AddConsumer {
+            consumer_address,
+            funds_available_for_staking,
+        } => sudo::add_consumer(deps, env, consumer_address, funds_available_for_staking),
+        SudoMsg::RemoveConsumer { consumer_address } => {
```

Comment:
no update_allowance? Or at least increase it?
(bonus point here)

Reply:
Will make a comment...